Mathlib/Mathport/Rename.lean

```diff
@@ -54,11 +54,16 @@ def RenameMap.insert (m : RenameMap) (e : NameEntry) : RenameMap :=
 /-- Look up a lean 4 name from the lean 3 name. Also return the `dubious` error message. -/
 def RenameMap.find? (m : RenameMap) : Name → Option (String × Name) := m.toLean4.find?
 
+-- TODO: upstream into core/std
+instance [Inhabited α] : Inhabited (Thunk α) where
+  default := .pure default
+
 /-- This extension stores the lookup data generated from `#align` commands. -/
-initialize renameExtension : SimplePersistentEnvExtension NameEntry RenameMap ←
+-- wrap state in `Thunk` as we rarely actually query it
```

Comment:
Isn't it queried every time that `#align` is used, to check for duplicates?

Reply:
How about we just rip out the duplication check at this point? It's not like we are adding new #align statements.

Reply:
Oof, not sure how I missed this. So this PR would "merely" reduce the latency up to the first `#align`, if any (which is true for most of Mathlib). I can think of solutions to remedy that but I'm not sure it's worth the complexity.

Reply:
@semorrison Ah, GitHub did not show me your message before. I'll leave that decision to you guys :) .

Reply:
I don't think removing the duplication check altogether is a good idea. We could consider only doing it if the `CI` variable is set, or as a liner pass?

Reply:
Given that:

> I made #8214 to compare with the result of just entirely removing `#align`. This PR picks up most of the gain already!

It sounds like we could just merge this PR (almost) as is anyway even though it's not as obvious *why* we get a performance improvement. If we do that, let's adjust this comment to not be misleading.

Reply:
Does this PR actually speed up the build time? What is the `open Mathlib` Benchmark testing?

This PR could be significant for downstream projects (that don't have any `#align` themselves).

Reply:
> I don't think removing the duplication check altogether is a good idea. We could consider only doing it if the `CI` variable is set, or as a linter pass?

What purpose is the duplication check serving? If we assume that no new `#align`s are being added, does it do anything?

Reply:
It prevents people blindly copying align statements, or making an accidental find and replace that touches the LHS of an align. I've had it catch mistakes multiple times.

Also, while outofsync is non-empty, new aligns are still being added.

Reply:
Okay.

I remain pretty skeptical that those things are worth on the order of `0.1s` of waiting time every time anyone opens a file in Mathlib.

Reply:
Where is that 0.1s estimate coming from? As far as I can tell neither this PR nor #8214 include a benchmark of the change you're proposing.

Reply:
My understanding is that the benchmark improvements seen above are real, but for now just deferred until the first #align. If the #align statements do not actually trigger this work (because they are not checking for duplicates and hence not hitting the lazy work set up in this PR), then that benchmark improvement turns into a real improvement for users.

Reply:
My understanding is that the PR is already an improvement for any user who runs `import Mathlib` and never writes `#align` (which is most non-mathlib users)